File: src/main/java/com/github/dkoval/leetcode/challenge/BestTimeToBuyAndSellStock3.java
package com.github.dkoval.leetcode.challenge;

/**
 * <a href="https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/">https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/</a>
 * <p>
 * You are given an array prices where prices[i] is the price of a given stock on the ith day.
 * <p>
 * Find the maximum profit you can achieve. You may complete at most two transactions.
 * <p>
 * Note: You may not engage in multiple transactions simultaneously (i.e., you must sell the stock before you buy again).
 */
public interface BestTimeToBuyAndSellStock3 {

    int maxProfit(int[] prices);

    // Resource: https://www.youtube.com/watch?v=0FKn0FSIQYE
    // O(N) time | O(N) space
    class BestTimeToBuyAndSellStock3DPBottomUp implements BestTimeToBuyAndSellStock3 {

        @Override
        public int maxProfit(int[] prices) {
            int n = prices.length;
            if (n < 2) {
                return 0;
            }

            // maxProfitLeft[i] is the maximum profit in prices[0:i] prefix
            int[] maxProfitLeft = new int[n];
            // the minimum price in prices[0:i] prefix
            int minPriceLeft = prices[0];

            // maxProfitRight[i] is the maximum profit in prices[i:] suffix
            int[] maxProfitRight = new int[n];
            // the maximum price in prices[i:] suffix
            int maxPriceRight = prices[n - 1];

            for (int i = 1; i < n; i++) {
                // iterate from 1 till n - 1
                maxProfitLeft[i] = Math.max(maxProfitLeft[i - 1], prices[i] - minPriceLeft);
                minPriceLeft = Math.min(minPriceLeft, prices[i]);

                // iterate from n - 2 down to 0
                int j = n - i - 1;
                maxProfitRight[j] = Math.max(maxProfitRight[j + 1], maxPriceRight - prices[j]);
                maxPriceRight = Math.max(maxPriceRight, prices[j]);
            }

            int maxProfit = 0;
            for (int i = 0; i < n; i++) {
                maxProfit = Math.max(maxProfit, maxProfitLeft[i] + maxProfitRight[i]);
            }
            return maxProfit;
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/BestTimeToBuyAndSellStock3.kt
package com.github.dkoval.leetcode.challenge

/**
 * [Best Time to Buy and Sell Stock III](https://leetcode.com/explore/challenge/card/august-leetcoding-challenge/551/week-3-august-15th-august-21st/3426/)
 *
 * Say you have an array for which the ith element is the price of a given stock on day i.
 *
 * Design an algorithm to find the maximum profit. You may complete at most two transactions.
 *
 * Note: You may not engage in multiple transactions at the same time (i.e., you must sell the stock before you buy again).
 */
object BestTimeToBuyAndSellStock3 {

    fun maxProfit(prices: IntArray): Int {
        var minPrice = Int.MAX_VALUE
        var maxProfitAfterFirstSell = 0
        var maxProfitAfterSecondBuy = Int.MIN_VALUE
        var maxProfitAfterSecondSell = 0
        for (price in prices) {
            minPrice = minOf(minPrice, price)
            maxProfitAfterFirstSell = maxOf(maxProfitAfterFirstSell, price - minPrice)
            maxProfitAfterSecondBuy = maxOf(maxProfitAfterSecondBuy, maxProfitAfterFirstSell - price)
            maxProfitAfterSecondSell = maxOf(maxProfitAfterSecondSell, price + maxProfitAfterSecondBuy)
        }
        return maxProfitAfterSecondSell
    }
}